### ISSUE TITLE
Use different quotation marks in echo, dump, write - solution found

### ISSUE THREAD

**Issue by siowyisheng** [kelaberetiv/TagUI#653]
`ERROR - [LINE 1] inconsistent quotes in "John's Bag"`

I would expect to be able to use single quotes within double quotes, but it errors out.

This is related/duplicate of #259 , but i think it's a significant shortcoming. Having to assign any text with quotes to a variable before echo is cumbersome and it's a gotcha for new users.

To see if there is a way to parse this properly in tagui

```
// to_replicate.tagui

echo "John's Bag"
```

**Comment by kensoh** [comment]
The fix for this is easy, and it's by design not a bug. In both tagui_parse.php and tagui_header.js there is a function add_concat(). That functions converts `'a' b 'c'` in TagUI into `'a' + b + 'c'` in JavaScript language for concatenation. You can look at the functions and without enforcing the quotes used in the string, there isn't a deterministic way to detect what user means in that string.

Original rationale was not to force users to use + + + to concatenate strings and variables, because that would be too programming-language-like vs human-language-like.

**Comment by siowyisheng** [comment]
> Without enforcing the quotes used in the string, there isn't a deterministic way to detect what user mean in that string.

I'm not sure if you're misunderstanding the complaint here. Users can't ever use single quotes in their strings when echoing (and dumping/writing).

**Comment by kensoh** [comment]
They can, just that they have to choose to use single quotes throughout and use the escape character like below -

```
echo '123\'456'
```

Or if want to use double quotes then use double quotes throughout -
```
echo "123\"456
```

Having a mix of both double and single quotes will make the add_concat() function unable to parse the string deterministically in the examples in my previous comment.

**Comment by siowyisheng** [comment]
```
echo 'John's Bag'
```
fails for me with `SyntaxError: Expected token ')'`

**Comment by siowyisheng** [comment]
```
echo 'John\'s Bag'
```
echoes `John's+Bag` for me. So it also doesn't work somehow.

**Comment by siowyisheng** [comment]
> Having a mix of both double and single quotes will make the add_concat() function unable to parse the string deterministically in the examples in my previous comment. 
I believe the first instance of `'` or `"` indicates which is being used as the surrounding quotes.

**Comment by kensoh** [comment]
By the time the string gets passed into the function add_concat(), the info that remains will be without the opening and closing quotes, so that info is no longer available to be used to decide which is the quotes used to denote the string.

Yes this is a bug, `echo 'John\'s Bag'` fails and `echo 'John\'sBag'` works. Looks like the function does not recognise the escaped ' and instead treat it like a delimiter, thus converts space to +.

**Comment by siowyisheng** [comment]
Ok will look through the relevant function and this bug.

**Comment by kensoh** [comment]
Sure! We'll can discuss more when we next meet.

**Comment by kensoh** [comment]
Besides fixing the bug, we could explore if there is a way to achieve the goal of the function while not restricting the use of 2 types of quotation marks. I didn't figure out a solution then, but now with 2 minds we might be able to introduce a good solution.

**Comment by kensoh** [comment]
Let me look into this. First on the bug with `echo 'John\'s Bag' and second on whether there is a viable implementation for using both type of quotation marks without parsing incorrectly.

**Comment by kensoh** [comment]
removing discussion label as this is a bug, and will review if there is a way to allow both quotes

**Comment by kensoh** [comment]
tagui_parse.php fix to output `echo 'John\'s Bag'` correctly. It works by introducing a new variable $previous_char to track what is the previous character. When the detected quote type is encountered and the previous character is the backslash, it will ignore the quotation mark and recognise that it is still within the string -
```php
function add_concat($source_string) { // parse string and add missing + concatenator
if ((strpos($source_string,"'")!==false) and (strpos($source_string,"\"")!==false))
{echo "ERROR - " . current_line() . " inconsistent quotes in " . $source_string . "\n";}
else if (strpos($source_string,"'")!==false) $quote_type = "'"; // derive quote type used
else if (strpos($source_string,"\"")!==false) $quote_type = "\""; else $quote_type = "none";
$within_quote = false; $source_string = trim($source_string); // trim for future proof
$previous_char = ""; // to help detect backslash escape for quotes
for ($srcpos=0; $srcpos<strlen($source_string); $srcpos++) {
if (($source_string[$srcpos] == $quote_type) and ($previous_char != "\\")) $within_quote = !$within_quote;
$previous_char = $source_string[$srcpos]; // to detect a previous backlash escape and ignore quote
if (($within_quote == false) and ($source_string[$srcpos]==" ")) $source_string[$srcpos] = "+";}
$source_string = str_replace("+++++","+",$source_string); $source_string = str_replace("++++","+",$source_string);
$source_string = str_replace("+++","+",$source_string); $source_string = str_replace("++","+",$source_string);
return $source_string;} // replacing multiple variations of + to handle user typos of double spaces etc
```

**Comment by kensoh** [comment]
js version for add_concat in tagui_header.js -
```js
function add_concat(source_string) { // parse string and add missing + concatenator
if ((source_string.indexOf("'") > -1) && (source_string.indexOf('"') > -1))
return "'ERROR - inconsistent quotes in text'";
else if (source_string.indexOf("'") > -1) var quote_type = "'"; // derive quote type used
else if (source_string.indexOf('"') > -1) var quote_type = '"'; else var quote_type = "none";
var within_quote = false; source_string = source_string.trim(); // trim for future proof
var previous_char = ""; // to help detect backslash escape for quotes
for (srcpos = 0; srcpos < source_string.length; srcpos++) {
if ((source_string.charAt(srcpos) == quote_type) && (previous_char !== "\\")) within_quote = !(within_quote);
previous_char = source_string.charAt(srcpos); // to detect a previous backlash escape and ignore quote
if ((within_quote == false) && (source_string.charAt(srcpos)==" "))
source_string = source_string.substring(0,srcpos) + "+" + source_string.substring(srcpos+1);}
source_string = source_string.replace(/\+\+\+\+\+/g,'+'); source_string = source_string.replace(/\+\+\+\+/g,'+');
source_string = source_string.replace(/\+\+\+/g,'+'); source_string = source_string.replace(/\+\+/g,'+');
return source_string;} // replacing multiple variations of + to handle user typos of double spaces etc
```

**Comment by kensoh** [comment]
Found solution to allow using both quotes, and your train of thought is correct, the first quotation mark encountered should be safely assumed to be the quotation mark chosen by user as the string enclosure quote. Fair expectation to have from user.

Below is the updated add_concat() for tagui_parse.php that works for the following script. I have to use 4 if else statement to do the check because comparing this condition `(strpos($source_string,"'") < strpos($source_string,"\""))` directly will lead to error when one of them is missing it will be comparing an integer to a false boolean and lead to wrong result. 

```
a = 'abc'
echo a
echo '1 2'
echo "1 2"
echo a " 1 ' \' \" 2 " a
echo a ' 1 " \" \' 2 ' a
echo 'a' "b" 'c' "d"
```
```
START - automation started - Sat Dec 28 2019 19:05:29 GMT+0800 (+08)

abc
1 2
1 2
abc 1 ' ' " 2 abc
abc 1 " " ' 2 abc
abcd

FINISH - automation finished - 0.2s
```
```php
function add_concat($source_string) { // parse string and add missing + concatenator
if ((strpos($source_string,"'") === false) and (strpos($source_string,"\"") === false)) $quote_type = "none";
else if ((strpos($source_string,"'") !== false) and (strpos($source_string,"\"") === false)) $quote_type = "'";
else if ((strpos($source_string,"'") === false) and (strpos($source_string,"\"") !== false)) $quote_type = "\"";
else if (strpos($source_string,"'") < strpos($source_string,"\"")) $quote_type = "'"; else $quote_type = "\"";
$within_quote = false; $source_string = trim($source_string); // trim for future proof
$previous_char = ""; // to help detect backslash escape for quotes
for ($srcpos=0; $srcpos<strlen($source_string); $srcpos++) {
if (($source_string[$srcpos] == $quote_type) and ($previous_char != "\\")) $within_quote = !$within_quote;
$previous_char = $source_string[$srcpos]; // to detect a previous backlash escape and ignore quote
if (($within_quote == false) and ($source_string[$srcpos] == " ")) $source_string[$srcpos] = "+";}
$source_string = str_replace("+++++","+",$source_string); $source_string = str_replace("++++","+",$source_string);
$source_string = str_replace("+++","+",$source_string); $source_string = str_replace("++","+",$source_string);
return $source_string;} // replacing multiple variations of + to handle user typos of double spaces etc
```

**Comment by kensoh** [comment]
add the matching add_concat() for tagui_header.js which also works for above tests.
```js
function add_concat(source_string) { // parse string and add missing + concatenator
if ((source_string.indexOf("'") == -1) && (source_string.indexOf('"') == -1)) var quote_type = "none";
else if ((source_string.indexOf("'") > -1) && (source_string.indexOf('"') == -1)) var quote_type = "'";
else if ((source_string.indexOf("'") == -1) && (source_string.indexOf('"') > -1)) var quote_type = '"';
else if (source_string.indexOf("'") < source_string.indexOf('"')) var quote_type = "'"; else var quote_type = '"';
var within_quote = false; source_string = source_string.trim(); // trim for future proof
var previous_char = ""; // to help detect backslash escape for quotes
for (srcpos = 0; srcpos < source_string.length; srcpos++) {
if ((source_string.charAt(srcpos) == quote_type) && (previous_char !== "\\")) within_quote = !(within_quote);
previous_char = source_string.charAt(srcpos); // to detect a previous backlash escape and ignore quote
if ((within_quote == false) && (source_string.charAt(srcpos) == " "))
source_string = source_string.substring(0,srcpos) + "+" + source_string.substring(srcpos+1);}
source_string = source_string.replace(/\+\+\+\+\+/g,'+'); source_string = source_string.replace(/\+\+\+\+/g,'+');
source_string = source_string.replace(/\+\+\+/g,'+'); source_string = source_string.replace(/\+\+/g,'+');
return source_string;} // replacing multiple variations of + to handle user typos of double spaces etc
```

**Comment by kensoh** [comment]
You can do some sanity checks and further tests to make sure this change doesn't break something or miss some use scenarios. After that you can merge as part of v6. Unassigning myself for now since no further work to be done from my side for now.

**Comment by kensoh** [comment]
Updating title, this issue is both a bug in existing implementation and a change from the existing design to support both types of quotation marks in steps like echo, dump, write, check..

**Comment by kensoh** [comment]
raised PR to merge into dev, unassigning myself

**Comment by siowyisheng** [comment]
fixed in #674 for v6 release
